### PR TITLE
core: place a cap on reorglogs

### PR DESCRIPTION
**Pull request by holiman**
This PR follows up on this comment:https://github.com/ethereum/go-ethereum/pull/24616/files#r839537977, since this did indeed cause an OOM situtation on this ticket: https://github.com/ethereum/go-ethereum/issues/25709#issuecomment-1239223539 . 

However, the OOM may have arisen due to other reasons, e.g. 32-bit arch being used.

### CODE REVIEW

**Comment by rjl493456442** [comment]
I am not sure it's a good idea to just throw the logs away. AFAIK, some applications will rely on these logs. Once the logs are removed because of reorg, they will have some logics to handle the reorg as well.

Maybe we can somehow split the logs to some batches? Just don't throw them away silently.

**Comment by MariusVanDerWijden** [comment]
Can't we just send the logs in batches?
e.g.
```diff
diff --git a/core/blockchain.go b/core/blockchain.go
index 506034b53..c918abe4a 100644
--- a/core/blockchain.go
+++ b/core/blockchain.go
@@ -2132,6 +2132,12 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
                if len(logs) > 0 {
                        rebirthLogs = append(rebirthLogs, logs)
                }
+               if i%10 == 0 {
+                       if len(rebirthLogs) > 0 {
+                               bc.logsFeed.Send(mergeLogs(rebirthLogs, false))
+                               rebirthLogs = rebirthLogs[:0]
+                       }
+               }
        }
        // If any logs need to be fired, do it now. In theory we could avoid creating
        // this goroutine if there are no events to fire, but realistcally that only
@@ -2140,9 +2146,6 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
        if len(deletedLogs) > 0 {
                bc.rmLogsFeed.Send(RemovedLogsEvent{mergeLogs(deletedLogs, true)})
        }
-       if len(rebirthLogs) > 0 {
-               bc.logsFeed.Send(mergeLogs(rebirthLogs, false))
-       }
        if len(oldChain) > 0 {
                for i := len(oldChain) - 1; i >= 0; i-- {
                        bc.chainSideFeed.Send(ChainSideEvent{Block: oldChain[i]})
```

**Comment by holiman** [comment]
I pushed a change now. The PR now does not discard any logs, however, I'm unsure if it fixed the problem. IMO it just shoves it down the line, hoping that the subscriptiion-system can handle unlimited number of blocks without causing OOM.

**Review by MariusVanDerWijden — APPROVED** [review]
LGTM, I think most of the time there is no one on the other side of the subscription, so pushing the issue further will actually solve the problem (in most cases^^)

File: core/blockchain.go [hunk 1]
```diff
@@ -2151,23 +2150,27 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
...
 	// networks where performance is not an issue either way.
 	if len(deletedLogs) > 0 {
 		bc.rmLogsFeed.Send(RemovedLogsEvent{mergeLogs(deletedLogs, true)})
 	}
+	// Collect the logs
+	var rebirthLogs []*types.Log
+	for i := len(newChain) - 1; i >= 1; i-- {
+		// Collect reborn logs due to chain reorg
+		if logs := bc.collectLogs(newChain[i].Hash(), false); len(logs) > 0 {
+			rebirthLogs = append(rebirthLogs, logs...)
```
Comment (holiman) [inline core/blockchain.go:2165]: Note for reviewer: I scrapped the append-then-use-`mergeLogs`, and just expanded the logs in-place immediately. I think it's done correctly

File: core/blockchain.go [hunk 2]
```diff
@@ -2151,23 +2150,27 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
...
+		}
+		if len(rebirthLogs) > 1000 {
+			bc.logsFeed.Send(rebirthLogs)
+			rebirthLogs = make([]*types.Log, 0)
```
Comment (fjl) [inline core/blockchain.go:2169]: Instead of this line, you could just as well reset the existing slice:

```
rebirthLogs = rebirthLogs[:0]
```

**Review by fjl — COMMENTED** [review]
The only consumer of this subscription is `filters.EventSystem` and it uses a channel buffer size of 10. (https://github.com/ethereum/go-ethereum/blob/d30e39b2f833fb75f1e529cd405061fb6b548b8d/eth/filters/filter_system.go#L143). 

So, whatever number you use as the 'batch size' for posting to the subscription, the maximum number of logs in memory is 10x that. Just something to keep in mind. Maybe it would make sense to limit the sent batches more aggressively here, like 512.

File: core/blockchain.go (hunk 1 above)
Comment (holiman) [reply core/blockchain.go:2165]: I have now fixed up the rebirth-log test, so I have now verified that it does indeed work as before

**Review by fjl — APPROVED** [review]

File: core/blockchain.go [hunk 3]
```diff
@@ -2151,24 +2150,31 @@ func (bc *BlockChain) reorg(oldBlock, newBlock *types.Block) error {
 		log.Crit("Failed to delete useless indexes", "err", err)
 	}
 
-	// Collect the logs
+	// Notify about removed logs first.      
+	if len(deletedLogs) > 0 {
```
Comment (rjl493456442) [inline core/blockchain.go:2154]: Can we do the same thing to `deletedLogs`?

It's same to `rebirthLogs`, the oldChain is also tracked, we can split the deletedLogs into different batches as well.

**Comment by rjl493456442** [comment]
General question. 

Previously there was a request to add a marker in `filterAPI` for indicating the log is the last one in a block. How did we implement it eventually? I remember we hold the assumption that for each log batch, it's the logs contained in a block or a few blocks. Splitting logs into batches randomly will lose this guarantee somehow?

**Comment by holiman** [comment]
The same guarantee holds in this PR. But no, we never did implement that ***@***.*** had an abandoned (?) PR about it iirc)

**Comment by holiman** [comment]
> had an abandoned (?) PR about 

This one: https://github.com/ethereum/go-ethereum/pull/24823

File: core/blockchain.go (hunk 3 above)
Comment (holiman) [reply core/blockchain.go:2141]: done

**Comment by fjl** [comment]
I think this PR is now absolutely perfect.

**Comment by rjl493456442** [comment]
> The same guarantee holds in this PR - we don't break randomly, we first add all logs from a block, then break

OK, then make sense!

**Review by rjl493456442 — APPROVED** [review]

**Pull request merged**